File: big-o-notation/exercises/two_sum/two_sum.py
"""
Problem: Two Sum
Complexity:
- Time: O(n)
- Space: O(n)

Memory Model:
- Python handles everything as an object in the Heap.
- 'nums' is a reference to a list.
- 'prevMap' is a reference to a dict.
- Variables like 'diff' or 'num' are references to immutable objects (integers).
"""

def two_sum(nums, target):
    prevMap = {} # val : index

    for i, n in enumerate(nums):
        diff = target - n
        if diff in prevMap:
            return [prevMap[diff], i]
        prevMap[n] = i
    return []

if __name__ == "__main__":
    nums = [2, 7, 11, 15]
    target = 9
    result = two_sum(nums, target)

    print(f"Input: nums = {nums}, target = {target}")
    print(f"Output: {result}")

    if result == [0, 1]:
        print("Test Passed! ✅")
    else:
        print("Test Failed! ❌")

File: big-o-notation/exercises/two_sum/two_sum.go
package main

import "fmt"

/**
 * Problem: Two Sum
 * Complexity:
 * - Time: O(n)
 * - Space: O(n)
 *
 * Memory Model:
 * - In Go, maps are reference types. When you pass a map to a function,
 *   you're passing a pointer to the underlying data structure in the Heap.
 * - Slices (like []int) are also reference-like, pointing to an array in the Heap.
 */

func twoSum(nums []int, target int) []int {
	prevMap := make(map[int]int)

	for i, num := range nums {
		diff := target - num
		if idx, ok := prevMap[diff]; ok {
			return []int{idx, i}
		}
		prevMap[num] = i
	}

	return nil
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9
	result := twoSum(nums, target)

	fmt.Printf("Input: nums = %v, target = %d\n", nums, target)
	fmt.Printf("Output: %v\n", result)

	if len(result) == 2 && result[0] == 0 && result[1] == 1 {
		fmt.Println("Test Passed! ✅")
	} else {
		fmt.Println("Test Failed! ❌")
	}
}
